revokefs/writer.c

```diff
@@ -0,0 +1,883 @@
+#include <sys/statvfs.h>
+#include <sys/socket.h>
+#include <stdio.h>
+#include <err.h>
+#include <stdlib.h>
+#include <string.h>
+#include <errno.h>
+#include <fcntl.h>
+#include <sys/xattr.h>
+#include <dirent.h>
+#include <unistd.h>
+#include <fuse.h>
+
+#include <glib.h>
+
+#include "writer.h"
+#include "libglnx.h"
+
+static int basefd = -1;
+
+static GHashTable *outstanding_fds;
+
+static GMutex mutex;
+
+static ssize_t
+do_request (int writer_socket,
+            RevokefsRequest *request,
+            const guchar *data,
+            size_t data_size,
+            const guchar *data2,
+            size_t data2_size,
+            RevokefsResponse *response,
+            guchar *response_data,
+            size_t response_data_size)
+{
+  size_t request_size;
+  size_t response_max_size;
+  ssize_t written_size, read_size, read_data_size;
+  struct iovec write_vecs[2] = {};
+  int n_write_vecs = 0;
+  struct iovec read_vecs[2] = {};
+  int n_read_vecs = 0;
+
+  request_size = sizeof (RevokefsRequest);
+  write_vecs[n_write_vecs].iov_base = (char *)request;
+  write_vecs[n_write_vecs++].iov_len = request_size;
+
+  if (data)
+    {
+      write_vecs[n_write_vecs].iov_base = (char *)data;
+      write_vecs[n_write_vecs++].iov_len = data_size;
+      request_size += data_size;
+    }
+
+  if (data2)
+    {
+      write_vecs[n_write_vecs].iov_base = (char *)data2;
+      write_vecs[n_write_vecs++].iov_len = data2_size;
+      request_size += data2_size;
+    }
+
+  written_size = TEMP_FAILURE_RETRY (writev (writer_socket, write_vecs, n_write_vecs));
+  if (written_size == -1)
+    {
+      g_printerr ("Write to socket returned error %d\n", errno);
+      return -1;
+    }
+  if (written_size != request_size)
+    {
+      g_printerr ("Partial Write to socket\n");
+      return -1;
+    }
+
+  response_max_size = sizeof (RevokefsResponse);
+  read_vecs[n_read_vecs].iov_base = (char *)response;
+  read_vecs[n_read_vecs++].iov_len = response_max_size;
+
+  if (response_data)
+    {
+      read_vecs[n_read_vecs].iov_base = response_data;
+      read_vecs[n_read_vecs++].iov_len = response_data_size;
+      response_max_size += response_data_size;
+    }
+
+  read_size = TEMP_FAILURE_RETRY (readv (writer_socket, read_vecs, n_read_vecs));
+  if (read_size == -1)
+    {
+      g_printerr ("Read from socket returned error %d\n", errno);
+      return -1;
+    }
+
+  if (read_size < sizeof (RevokefsResponse))
+    {
+      g_printerr ("Invalid read size %zd\n", read_size);
+      return -1;
+    }
+
+  return read_size - sizeof (RevokefsResponse);
+}
+
+static int
+request_path_i64_i64 (int writer_socket, RevokefsOps op, const char *path, guint64 arg1, guint64 arg2)
+{
+  RevokefsRequest request = { op };
+  RevokefsResponse response;
+  size_t path_len = strlen (path);
+  ssize_t response_data_len;
+
+  if (path_len > MAX_DATA_SIZE)
+    return -ENAMETOOLONG;
+
+  request.arg1 = arg1;
+  request.arg2 = arg2;
+
+  g_mutex_lock (&mutex);
```

Comment:
Why not lock & unlock the mutex inside `do_request` instead? That seems much cleaner that having to remember to lock & unlock every time it's called.